### PR TITLE
Build libcontainer's container on Instance::new to be spec compliant

### LINKED ISSUE #158: Instance::new should return a Result type
This was discussed before: https://github.com/containerd/runwasi/pull/54#issuecomment-1403269766

IMHO we should aim to remove as many `unwrap()` calls as possible from shim "main thread", because a library should not panic easily.

_Originally posted by @ipuustin in https://github.com/containerd/runwasi/pull/142#discussion_r1238411469_

### PR DESCRIPTION
**Pull request by jprendes**
Fixes #158 and #326 

The issue behind #326 is that we are asking youki to **create** and **start** the container on the **start** call.
We don't do any meaningful work on the **create** call.

This missmach means that containerd is assuming that extra setup has happened successfully. but youki hasn't done it yet. When youki fails to do it at **start**, we are breaking those assumptions (e.g., being able to provide a PID for the init process).
The solution is to `build` (**create**) the youki container on the create call (i.e., the `new` method of runwasi's instance).
This also requires that we change the signature of `new` to return a `Result<Self, Error>`, which was previously mentioned in #158.

~~I haven't added any test for this here as testing it is rather involved.
The sideeffect of this issue on the first invocation is that we leave a dangling process, and on a second invocation is a hang.
I'm open to suggestions on how we could test it.~~ Thanks @jsturtevant for suggesting testing approach.

### CODE REVIEW

**Comment by jprendes** [comment]
> this is the side affect of this change?

No, this is what I observe on linux before this change. With this change I do not observe any dangling process and no hanging.

**Comment by jprendes** [comment]
> Would it be something like:

Yes, that's 100% correct, thanks!. I'll add test points for all shims.

**Comment by jprendes** [comment]
I gave the test a go. The right place for it would be in the `containerd-shim-wasm` crate, as there's where the logic is implemented.

However, that test uses the `containerd-shim-wasm-test` crate. Adding that crate as a dependency creates a cycle which causes some weird interaction with types being doubly defined (e.g., there's a `crate::sandbox::Instance` and a `containerd_shim_wasm::sandbox::Instance`, and they are not the same trait for the compiler)

A solution is bringing the testing to the `containerd-shim-wasm-test` functionality to the `containerd-shim-wasm` behind a feature flag. Doing that significantly increases the number of files touched by this PR, but shouldn't add much complexity.

**Comment by jprendes** [comment]
I recommend reviewing the commits separately :-)

File: crates/containerd-shim-wasm/Cargo.toml
Comment (jsturtevant) [inline crates/containerd-shim-wasm/Cargo.toml:21]: since these are under dev-dependencies are they needed here?

File: crates/containerd-shim-wasm/src/sys/unix/container/instance.rs
Comment (jsturtevant) [inline crates/containerd-shim-wasm/src/sys/unix/container/instance.rs:121]: are we losing the fact that it might fail here?

File: crates/containerd-shim-wasm/src/sys/unix/container/instance.rs
Comment (jsturtevant) [inline crates/containerd-shim-wasm/src/sys/unix/container/instance.rs:38]: since we return an error here, we can bubble this error up now?

File: crates/containerd-shim-wasm/src/sys/unix/container/instance.rs
Comment (jsturtevant) [inline crates/containerd-shim-wasm/src/sys/unix/container/instance.rs:33]: since this will panic as well we could clean these up while we are here.

**Review by jsturtevant — COMMENTED** [review]
great find!  Creating the container during the new aligns with containerd's "task_create" which makes sense. left a few minor comments

File: crates/containerd-shim-wasm/Cargo.toml
Comment (jprendes) [reply crates/containerd-shim-wasm/Cargo.toml:21]: For consumers of the `testing` feature of the library we need to specify it as a normal dependency as well.
All 3 dev-dependencies are also specified as optional normal dependencies.
I'm open to suggestions :smile:

File: crates/containerd-shim-wasm/src/sys/unix/container/instance.rs
Comment (jprendes) [reply crates/containerd-shim-wasm/src/sys/unix/container/instance.rs:121]: Indeed, I missed that.

File: crates/containerd-shim-wasm/src/sandbox/shim.rs
Comment (jsturtevant) [inline crates/containerd-shim-wasm/src/sandbox/shim.rs:749]: this is safe since `Nop::new()` always returns `Ok`.  might be work a comment incase things change in future?

**Review by jsturtevant — COMMENTED** [review]
LGTM

**Comment by devigned** [comment]
@jprendes this needs a rebase, but lgtm after that.

**Review by devigned — APPROVED** [review]
lgtm!

**Pull request merged**
